### ISSUE TITLE
Windows 10 2004-20H2 Xinput devices randomly not being released upon disconnect

### ISSUE THREAD

**Issue by DaRkL3AD3R** [Ryochan7/DS4Windows#1597]
**Describe the bug**
On a fresh boot and launch of DS4Windows, connecting the controller (either USB or BT) will result in the first virtual Xbox controller being mapped to device slot 0 (first available). Sometimes after exiting and disconnecting from DS4Windows, this virtual controller will remain attached and applications that expect to only use this first virtual controller will ignore any subsequent virtual controllers spawned by DS4Windows, which end up occupying the next available slot, slot 1 (second). The only way to resolve this problem until it happens again, is to reboot the PC.

**To Reproduce**
Unknown exact trigger but launching DS4Windows, connecting a controller, then closing the application and disconnecting the controller seems to be the general method to causing it to happen.

**Expected behavior**
Prior to Windows 2004, DS4Windows would always maintain the first available Xinput device slot no matter how many times you open and close, connect and disconnect controllers to the virtual driver, never causing any conflicts with input being ignored in various apps and games.

**Screenshots**
https://i.imgur.com/hPauaBo.png
In the above screenshot of the Dolphin controller input device menu, you can see that there is a [disconnected] virtual Xbox 360 controller in slot 1 from a prior launch. This launch, DS4Windows is hooking to Xinput slot 0 properly.

**Desktop (please complete the following information):**
 - Controller Make and Model: Sony DualSense
 - OS: Windows 10 Pro Version 20H2 Build 19042.630
 - DS4Windows Version: 2.1.17

**Comment by DaRkL3AD3R** [comment]
I'm pretty sure the step to reproduce this is simply to stop and start the controller service inside DS4Windows. Closing the application completely and opening it again will have the same effect. That Xinput device 0 slot just gets stuck as present but "disconnected" to the system and the next available slot is used.

**Comment by mika-n** [comment]
This is an old random issue and the problem is not in DS4Windows app itself, but in ViGem driver. Unfortunately DS4Windows app cannot do much about it if this "ghost device" issue happens. The exact scenario is still a bit unclear why it happens in some PCs (for example I have never experienced it).

Anyway, the "persistent output slot" feature may help to tackle this issue. If you set a slot as "Persistent" instead of Dynamic in OutputSlots tab page then DS4Windows keeps the virtual output controller in connected state (as long DS4Windows app is running) even when the physical gamepad is disconnected. The next time you connect the gamepad it re-uses the existing persistent virtual output controller. This probably helps to work around the ghost controller issue because virtual output devices are not constantly created and destroyed when a physical game is connected/disconnected.

**Comment by DaRkL3AD3R** [comment]
Yeah I've been using a permanent persistent virtual controller and making DS4Windows run on startup. It's weird for me because I have to change my mentality of how I use the program. I went from closing DS4Windows to power off my controller to now I have to keep it running the whole boot cycle or else I get hit with the glitch. At least it's a decent enough workaround for the time being. Hopefully it can get sorted. I'm guessing it has something to do with Windows 10 2004+ builds as I've been using this controller wrapper for years without this issue.

**Comment by Ryochan7** [comment]
Been testing this type of scenario with ViGEmBus 1.17.333.0 and I have not run into this problem yet. The latest ViGEmBus driver has fixed a couple of other issues that I have encountered with previous releases. I made a small test program when testing the initial migration and even something like accessing the User index of a controller was extremely flawed with the 1.16.112.0 driver.

https://gitlab.com/ryochan7/vigemplugtest

https://github.com/ViGEm/ViGEmBus/releases/tag/setup-v1.17.333

**Comment by Ryusennin** [comment]
Indeed, ViGEm 1.17 seems to have fixed the issue for me. I tried quitting/restarting DS4Win several times and XInputChecker confirms that Controller #1 always reconnects to the first available slot.

**Comment by DaRkL3AD3R** [comment]
I can also confirm that after updating to the new version with ViGEmBus update, the issue no longer presents itself. Thanks for letting me know. Won't have to have a permanent X360 device running at startup anymore. Cheers!

**Comment by DaRkL3AD3R** [comment]
I knew it I spoke too soon. Issue happened again just now. I connected the controller (a DS5) and it didn't get exclusive access to it. Then when I disconnected it, it kept the Xinput Device #0 and I had to reboot my PC to get control to that slot again. Welp looks like back to having the permanent device added to the output slot and having it run on bootup. Shame, really thought it was fixed.

**Comment by Ryusennin** [comment]
It looks like you are right. I did a new series of tests today and slot 0 is unavailable after simply quitting and relaunching DS4Win. It's strange as it really worked nicely yesterday.

**Comment by DaRkL3AD3R** [comment]
Has anyone managed to find a way to delete the virtual controller from existence during run-time? I refuse to believe there is no way to alleviate that slot reservation without rebooting. And furthermore I find it rather odd that there's no mention of this issue on the ViGEm github issues list. This is a pretty big problem, considering how stable everything else is. Using a permanent slot reservation doesn't really solve it when you want to update DS4Windows on launch. That's all it takes to trigger the bug.

**Comment by Ryochan7** [comment]
> Using a permanent slot reservation doesn't really solve it when you want to update DS4Windows on launch. That's all it takes to trigger the bug.

Even that doesn't make it happen on a Windows 7 VM for me. And not experiencing it myself is the biggest problem. I pretty much need someone who is experiencing the problem to debug it since I cannot reproduce it at all with the latest ViGEmBus driver.

**Comment by DaRkL3AD3R** [comment]
What version of 10 are you running? I only started having this problem on 2004 and up.

**Comment by Ryochan7** [comment]
I am currently running Windows 10 Release ID 2004

**Comment by Ryochan7** [comment]
Even with the older ViGEmBus driver, the problem would occur so rarely that I could not debug it. The problem had popped up in my alternative mapper programs as well so it seemed to be an issue beyond DS4Windows and its parts.

**Comment by DaRkL3AD3R** [comment]
It's definitely a vigembus issue not DS4Windows. It's just a shame that it's clearly an issue people are facing and it seems pretty ignored. I'm not trying to ride you here ryochan you've been really open and speculative about it and I agree it's not your software at fault. I just lack the chops to make any headway on debugging it which is a shame because I can reproduce it on command every boot simply by starting and stopping the service with a permanent virtual 360 controller plugged in.

**Comment by mika-n** [comment]
Have you disabled the "Fast startup" crap in Win10? Especially nowadays when most people have (or at least should have) fast SSD disks instead of ultra-slow HDD disk the fast startup cache is useless and causes just ghost bugs over WinOS restarts.
https://www.windowscentral.com/how-disable-windows-10-fast-startup

If you can describe and create a re-producable usage pattern then you might wanna try to reach out to ViGem support forum. But then again if devs (ds4win or vigem devs) are not able to re-produce it then it is veeeeeery difficult to track down (some PC specific/setup/driver/app combination clitch).

**Comment by Ryochan7** [comment]
**TLDR: Had the problem occur last night. The only recourse seems to be a Windows reboot.**

Had the issue pop up last night using ViGEmBus 1.17.333.0. First noticed an issue when trying to load Metal Gear Rising: Revengeance as the game hard locked on startup. DS4Windows failed to connect to ViGEmBus after that point as the virtual device plugin routine would constantly fail and proceed to close the connection to ViGEmBus. XInputChecker showed that XInput Controller slot 1 was still seen as connected.

Some other experiments with Device Manager didn't help. The DInput interface HID controller was in an invalid state. Neither the ViGEmBus device nor the corrupt HID controller could be disabled. The only recourse was to reboot Windows. Of course, that reboot resulted in a BSOD. Here is a screenshot showing Device Manager in that state.

![xinput_controller_broken](https://user-images.githubusercontent.com/581968/107984782-4c3c6e00-6f8e-11eb-9634-93345716801b.png)

At least now, it is not just a freak coincidence but a real indicator of driver failure; maybe not even in ViGEmBus directly but an unclean close from the XUSB driver. I cannot remember when I put in the virtual plugin device check in DS4Windows but it helped with discovering a problem. Closing the connection to ViGEmBus seems to be the best route to take if the problem ever occurs. I might want to change the log message that pops up when the virtual plugin routine fails and increase its severity.

**Comment by FJRMaverick** [comment]
Confirming the general problem with DS4Windows v3.06 (still there because of other problems with later versions) and
Edition	Windows 10 Pro
Version	21H1
Installed on	‎14/‎09/‎2020
OS build	19043.1052
Experience	Windows Feature Experience Pack 120.2212.2020.0

It is not happening, when switching the controller off and on again, but only when stopping/restarting the service within DS4Windows or ending and starting the whole program.

Removing the controller from windows via "Settings / Devices" and new paring/connecting resolved the problem **without reboot**

Also setting "Disconnect from BT when stopping" works as a workaround, the issue is not happening then, the output slot is freed when stopping DS4Windows or finishing/restarting it, though ofc you have to switch on the controller afterwards.

**Comment by DaRkL3AD3R** [comment]
I'm not entirely sure about that setting, disconnect from BT when stopping. I have had that set since day one and the problem still occurs when I exit the program/stop the service. As for turning the controller on and off, last time intestedi tested it around a week ago with dynamic hotplugging, the first slot still became broken.

**Comment by FJRMaverick** [comment]
Dunno, yet I can only talk about my experience. And what I wrote above was tested today with the mentioned results.

**Comment by Ryusennin** [comment]
Disconnect from BT doesn't do a thing for me either, and the problem remains for wired connections.

My best workaround for now is to log off and on again. Faster than a reboot, although of course you need to abandon your current Windows session and that can be a hassle.

**Comment by Ryochan7** [comment]
This seems to be confusing two different problems.

> It is not happening, when switching the controller off and on again, but only when stopping/restarting the service within DS4Windows or ending and starting the whole program.

I am going to assume you have customized the Output Slots and set some controllers as Permanent. An issue with Permanent slots sometimes not getting detached and output controllers unplugged upon Service Stop was fixed with version 3.0.7; it was a long standing issue that did not have a definitive way to test myself until recently. That was entirely a DS4Windows problem. 99% of the time, I just use Dynamic slots so I would expect issues raised regarding skipped XInput slots to be for normal virtual controller unplugging; I was not told otherwise.

The issue I brought up in my last comment in this thread has been somewhat addressed. The new PurgeOldXInput tool would need to be run to purge all associated XInput controllers from the system. That way, DS4Windows would no longer immediately disconnect from ViGEmBus after a device plugin failure. Luckily, I have no run into that issue again but at least there is a way around the problem now.

**Comment by DaRkL3AD3R** [comment]
Okay so based on your comment Ryochan I just did a bunch of tests as of version 3.0.9.

- Having no permanent slots, only dynamic slots, works flawlessly. Can connect/disconnect controller, can start/stop service, can open/close DS4Windows and the controller always assigns to Xinput Slot #0.

- Initially I had DS4Windows set to run on startup and automatically setup a permanent Xbox controller which worked flawlessly when connecting/disconnecting the real controller, however as soon as I changed it to dynamic or stopped the service/closed DS4Windows, subsequent controller connections would spawn on slot #1. This just happened before as I read your reply. Then I signed out and back in, reconfigured the controller as dynamic, and no matter how I set it back to permanent and test it, it never assigns to slot #1 anymore. It's always the proper #0 on every connection.

So as of this moment on v3.0.9, everything seems to be working as it should. I guess I'll keep an eye out for that strange behavior and if in the future it ends up happening again, I'll report back to this thread. For the time being though it truly does seem to be fixed this time. Amen.
